feat: jet 部署 helm chart

YkeLit opened the pull request:
1. deployment 与 service 资源清单
  - project man
  - trace aware
  - breeze
2. ingress route 入口路由
  - breeze
  - project man
3. middleware
  - compress
  - redirect-to-https
  - strip-prefix
  - cors
4. dependencies
  - bitnami postgresql

YkeLit committed:
feat: add helper templates

define name and full name
define labels

YkeLit committed:
feat: add deployment template

define image name and image pull secret
add deployment template and values

YkeLit committed:
feat: add image pull secret template

YkeLit committed:
feat: add service template

YkeLit committed:
feat: add jet env configmap template

YkeLit committed:
feat: add jet ingress template

 support multiple hosts and path
 supports certificate challenge or custom certificate

YkeLit committed:
feat: add certificate secret template

YkeLit committed:
feat: add traefik middleware template

add compress, strip sub path and redirect to https middlewares

YkeLit committed:
feat: add chart default files

YkeLit committed:
feat: add velero backup source template

YkeLit committed:
feat: add cors setting

YkeLit committed:
test: add chart unit test

test deployment, secrets, ingressroute and  middleware

YkeLit committed:
feat: add breeze source list

add deployment and service source list

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,104 @@
+env:
+  RELEASE_NODE: sname
+  RELEASE_COOKIE: cookie
+  CREDENTIAL_SECRET: gBgN6F4183LXKFbiE59K/zq+VuX1m3x2cf1lry+TEdflE3i4IjUYGObOlmyKm2Op
```
不给密钥
```suggestion
  CREDENTIAL_SECRET: my-credential-secret
```

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,104 @@
+  PROJECT_MAN_GRPC_SERVER_PORT: 50051
+  JET_API_ENDPOINT: http://jet/api/plugin
+  JET_JWT_PRIVATE_KEY: ""
+  SECRET_KEY_BASE: Sl7ym/kDmxL7mcZzkY7S/zLEP9Ybkp4aqTnE/BoHQfsGAwnKqQaKfyzWPs4pAebU
```
```suggestion
  SECRET_KEY_BASE: my-secret-key-base
```

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,104 @@
+
+breeze:
+  image:
+    repository: jet
```
```suggestion
    repository: breeze
```

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,104 @@
+  # - /api
+  # - /plugin
+  # - /plugin/explorer
+  # - /plugin/graphql.sdl
```
这个为什么做成可以配置的呢？

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,104 @@
+  # storageLocation: ""
+  # ttl: "720h0m0s"
+
+corsSettings: {}
+  # accessControlAllowHeaders:
+  #   - '*'
+  # accessControlAllowMethods:
+  #   - GET
+  #   - OPTIONS
+  #   - PUT
+  # accessControlAllowOriginList:
+  #   - '*'
+  # accessControlMaxAge: "100"
```
middleware 放一起吧

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,104 @@
+  # - /plugin/explorer
+  # - /plugin/graphql.sdl
+jetStripPrefix: ""
+jetTLS: {}
+  # Certificate challenge field
+  # certResolver: jet
+  #
+  # Custom tls certificate
+  # tlsSecret: {}
+    # certificate: base64 encoding certificate-file
+    # key: base64 encoding key-file
```
这个是不是用 k8s 的 secret 来管理

fahchen commented on templates/secrets/docker-registry.yaml:
这个是不是我们引用 k8s 的一个 secret，而不是我们自己创建一个？

fahchen commented on templates/configmap/jet.yaml:
```diff
@@ -0,0 +1,9 @@
+metadata:
+  labels:
+    {{- include "jet-helm-chart.labels" . | nindent 4 }}
+  name: {{ .Chart.Name }}-env
```
name 应该不对，不是 chart name

fahchen commented on templates/deployment.yaml:
```diff
@@ -0,0 +1,43 @@
+      imagePullSecrets:
+        - name: {{ include "jet-helm-chart.fullname" . }}-docker-registry-secret
+      containers:
+        - name: {{ .Chart.Name }}
```
这个为什么直接事 chart name 呢？
```suggestion
        - name: {{ include "jet-helm-chart.image" . }}-jet
```

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,104 @@
+affinity: {}
+
+# Jet environment variable
+env:
```
这里是不是把 breeze，jet，trace_aware 的 env 分开呀？

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,104 @@
+  # - /plugin
+  # - /plugin/explorer
+  # - /plugin/graphql.sdl
+jetStripPrefix: ""
```
这里设计成一个变量 `subpath` 吧

1. https://github.com/rails/rails/blob/main/guides/source/configuring.md#configrelative_url_root
2. https://hexdocs.pm/phoenix/Phoenix.Endpoint.html#module-runtime-configuration

fahchen reviewed (commented):
这个 helm-chart 我们有几个服务
1. project_man
2. trace_aware
3. breeze
4. 后面可能还有其他

所以这个 chart 里面的这些变量参数注意区分一下

YkeLit committed:
feat: add project man deployment and ingress resoure list

use exist image pull secret or create image pull secret

YkeLit committed:
test: add project man and change middleware test

YkeLit committed:
feat: add project man deployment values

YkeLit committed:
feat: support exist image pull secrets

support exist image pull secrets
also can create image pull secrets

YkeLit committed:
feat: add middlewares field and subpath function

subpath and cors in middlewares field
add subpath helper function

YkeLit committed:
refactor: breeze source list change

add labels helper
change deployment image pull secrets
add ingress route source
change project man default service port

YkeLit committed:
test: add breeze unit test

YkeLit committed:
feat: add trace aware source list

add helper file
add deployment and service source

YkeLit committed:
test: add trace aware unit test

add deployment test

YkeLit committed:
chore: add helm chart ci

YkeLit committed:
docs: add chart instructions

YkeLit committed:
refactor: rename trace aware file name

YkeLit committed:
feat: add db dependencies

add jet and dynamic postgres chart
add subcharts values

YkeLit committed:
feat: add initialization check

YkeLit commented:
@fahchen 飞哥有空看一下

fahchen commented on Chart.yaml:
```diff
@@ -0,0 +1,34 @@
+  - name: postgresql
+    version: 12.5.7
+    repository: https://charts.bitnami.com/bitnami
+    alias: dynamicdb
```
这个名字我想还是调整一下叫做 projectsdb
@vanppo

vanppo replied:
Jet 代码要调吗

fahchen replied:
确定的话就要一起调相关的命名

vanppo replied:
就是 `ProjectMan.Repo.Dynamic` 改成 `ProjectMan.Repo.Project` 嘛 我觉得都可以 看你要不要改 dynamic 这个命名是从 Ecto 那边来的

YkeLit replied:
要改的话 env 也要改吧

fahchen commented on README.md:
```diff
@@ -0,0 +1,96 @@
+| middlewares.corsSettings | object | `{}` | 跨域配置 |
+| middlewares.jetSubpath | string | `""` | jet 子路径 |
+| backup | object | `{}` | 数据库备份设置 |
+| jetdb.auth.postgresPassword | string | `"changeit"` | postgres 用户密码设置 |
```
这个能做到 secret 里面不呀，其它的 password 也一样

YkeLit replied:
secret 资源来存这些密码哇，单独设置 secrets 一级

fahchen replied:
对

fahchen commented on README.md:
```diff
@@ -0,0 +1,96 @@
+| env.TRACE_AWARE_DATABASE_URL | string | `"ecto://postgres:changeit@jet-postgresql/trace_aware_prod"` | Trace Aware 数据库连接地址（格式：`ecto://USER:PASS@HOST/DATABASE`，[参考](https://hexdocs.pm/ecto/3.9.4/Ecto.Repo.html#module-urls)） |
+| env.TRACE_AWARE_GRPC_SERVER_PORT | string | `"80"` | Trace Aware GRPC 服务监听的端口号 |
+| jetHosts | list | `[]` | jet 的访问地址 |
+| jetTLS | object | `{}` | tls 证书 |
```
这个也做的 secret 里面一下吧

YkeLit replied:
这个是 secret 资源的嘛

fahchen replied:
现在在 values里面存放的是证书，要改成 secret 的名字，secret 单独创建，不在这个 chart 中维护

fahchen commented on templates/_helpers.tpl:
```diff
@@ -0,0 +1,82 @@
+{{/*
+docker registry auth credenttial
+*/}}
+{{- define "jet-helm-chart.imagePullSecret" }}
```
这个走 secret 呢

YkeLit replied:
这个也是 secret

YkeLit replied:
这个我觉得可以留着，这个应该被 helm 统一管理起来，可以加一个调用已存在 secret 的参数。

fahchen commented on templates/_helpers.tpl:
```diff
@@ -0,0 +1,82 @@
+{{/*
+Host for access rule
+*/}}
+{{- define "jet-helm-chart.jetHosts" -}}
```
这个 helper 的命名不太贴切

fahchen commented on templates/_helpers.tpl:
```diff
@@ -0,0 +1,82 @@
+{{/*
+Selector labels
+*/}}
+{{- define "jet-helm-chart.selectorLabels" -}}
```
selector-labels 还是小驼峰呢，看看有没规范，统一一下

YkeLit replied:
https://helm.sh/zh/docs/chart_best_practices/templates/#%E5%AE%9A%E4%B9%89%E6%A8%A1%E6%9D%BF%E7%9A%84%E5%90%8D%E7%A7%B0
官方并没有明确规范模版名称，但是有个建议，这个是通过 `helm create` 创建出来的模版之一。命名的话更推荐小驼峰吧。

fahchen replied:
就统一成小驼峰吧

fahchen replied:
看下能不能加个 lint 检测下

YkeLit replied:
只有手写一个，没有找到现成的 lint 工具

fahchen commented on templates/breeze/_helpers.tpl:
```diff
@@ -0,0 +1,29 @@
+{{- define "jet-helm-chart.breeze-labels" -}}
+helm.sh/chart: {{ include "jet-helm-chart.chart" . }}
+{{ include "jet-helm-chart.breeze-selectorLabels" . }}
+{{- if .Chart.AppVersion }}
```
这个条件判断有什么意图呀

YkeLit replied:
https://helm.sh/zh/docs/topics/charts/#chartyaml-%E6%96%87%E4%BB%B6
这个字段是可选字段，所以加了个判断避免报错

fahchen replied:
对于我们这个具体的 chart 也是吗

YkeLit replied:
这个具体的 chart 是什么意思呀

fahchen replied:
就是对于 jet-helm-chart 这个字段是不是可选的

fahchen commented on templates/breeze/deployment.yaml:
```diff
@@ -0,0 +1,57 @@
+          args:
+            - 'start'
+            - '-j'
+            - 'http://{{ include "jet-helm-chart.fullname" . }}-project-man:{{ .Values.projectMan.service.prot }}'
```
```suggestion
            - 'http://{{ include "jet-helm-chart.fullname" . }}-project-man:{{ .Values.projectMan.service.port }}'
```

fahchen commented on templates/breeze/deployment.yaml:
```diff
@@ -0,0 +1,57 @@
+            - '0.0.0.0:{{ .Values.breeze.service.port }}'
+          envFrom:
+            - configMapRef:
+                name: {{ include "jet-helm-chart.fullname" . }}-env
```
这个分开写一下，breeze 用自己的 env（内容现在可以都一样，开发进来调整）

fahchen commented on templates/breeze/ingressroute.yaml:
```diff
@@ -0,0 +1,65 @@
+{{- end }}
+{{- $webMiddlewares = append $webMiddlewares "compress" }}
+{{- if .Values.jetTLS }}
+{{- $webMiddlewares = append $webMiddlewares "redirect-to-https" }}
```
这个有没可能在 ingress route 层面做一次呢，而不是每个都做一下（比如 breeze，project_man 都做了一次）

YkeLit replied:
traefik 有个 `--entrypoints.<name>.http.redirections.entrypoint.to=websecure` 可以设置
https://doc.traefik.io/traefik/reference/static-configuration/cli/

fahchen commented on templates/projectman/_helpers.tpl:
```diff
@@ -0,0 +1,29 @@
+{{/*
+Return the jet image name
+*/}}
+{{- define "jet-helm-chart.projectMan-image" }}
```
```suggestion
{{- define "jet-helm-chart.project-man-image" }}
```

fahchen commented on templates/projectman/_helpers.tpl:
```diff
@@ -0,0 +1,29 @@
+*/}}
+{{- define "jet-helm-chart.projectMan-selectorLabels" -}}
+{{ include "jet-helm-chart.selectorLabels" . }}
+app.kubernetes.io/application: projectMan
```
```suggestion
app.kubernetes.io/application: project-man
```

fahchen commented on templates/projectman/service.yaml:
文件路径
projectman -> project-man

fahchen commented on templates/traceaware/_helpers.tpl:
```diff
@@ -0,0 +1,29 @@
+{{/*
+Return the jet image name
```
注释和实际内容不符合

fahchen commented on templates/traceaware/deployment.yaml:
```diff
@@ -0,0 +1,59 @@
+          command:
+            - 'sh'
+            - '-c'
+            - 'until psql -c "SELECT 1" "user=postgres password={{ .Values.jetdb.auth.postgresPassword }} dbname=trace_aware_prod host={{ .Values.jetdb.fullnameOverride }}"; do echo waiting for trace aware db; sleep 2; done'
```
trace-aware 和 project-man 用的是一个数据库实例吗？
@vanppo

vanppo replied:
当时设计的时候是用同一个的 这样 project_man 能直接从数据库里面读日志 否则只能通过 trace_aware 的接口来取

fahchen replied:
现在这样会导致数据库连接池加倍的哦，得分开做成两个数据库连接

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,224 @@
+  TRACE_AWARE_GRPC_SERVER_PORT: "80"
+
+# Jet ingress route
+jetHosts: []
```
不用带 jet 前缀

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,224 @@
+
+# ingress middlewares
+middlewares:
+  jetSubpath: ""
```
这个提升到 host 那一层吧
```suggestion
  subpath: ""
```

fahchen commented:
别 force push 哈

YkeLit committed:
refactor: change helper function name

jetHosts --> jetRuleHosts

YkeLit committed:
fix: The comments do not match the actual content

YkeLit committed:
refactor: Change project man file name

YkeLit committed:
refactor: Change trace aware file name

YkeLit committed:
refactor: change jetHosts to hosts and helper function jetRuleHosts to ruleHosts

YkeLit committed:
test: change values jetHosts name

YkeLit committed:
fix: deployment spec.template.spec.containers[0].args values name error

YkeLit committed:
style: template name follow the Camel-Case format

YkeLit committed:
style: project man template name follow the Camel-Case format

YkeLit committed:
style: breeze template name follow the Camel-Case format

YkeLit committed:
style: trace aware template name follow the Camel-Case format

YkeLit committed:
refactor: change trace aware labels application name

YkeLit committed:
refactor: change project man labels application name

YkeLit committed:
refactor: change jetSubPath in root field

YkeLit committed:
feat: add breeze env field

YkeLit committed:
feat: remove tls source
Manual creation tls secret source

YkeLit committed:
feat: Add the secret field to store password sensitive information

YkeLit committed:
docs: Get the  access address

YkeLit committed:
refactor: change Values jetSubpath to subpath

YkeLit committed:
refactor: Delete redundant environment variables

YkeLit committed:
docs: add chart deploy values guide

fahchen commented on README.md:
```diff
@@ -0,0 +1,304 @@
+| fullnameOverride | string | `""` |  |
+| nameOverride | string | `""` |  |
+| replicaCount | int | `1` |  |
+| resources | object | `{}` |  |
```
这两个应该要每个 deployment 单独控制

YkeLit commented on .github/workflows/lint-test.yaml:
缺少单元测试

YkeLit committed:
refactor: change jetdb name to projectmandb

YkeLit committed:
refactor: remove imagePullSecre flied

manual management imagePullSecret

YkeLit committed:
refactor: deployment templates to use specific resources, nodeSelector, tolerations, and affinity for each microservice.

YkeLit committed:
refactor: add project man env

YkeLit committed:
refactor: init container use exist secret

YkeLit committed:
refactor: Update replica count in deployment.yaml templates.

YkeLit committed:
refactor: 调整 jet 各个应用环境变量

通过 tpl 方法使 values 文件中 service 相关变量关联
将某些固定变量直接写入 configMap 资源清单中
添加构建 jet api endpoint 与 breeze address 方法

YkeLit committed:
refactor: Use configMapRef for environment variables in breeze deployment

YkeLit committed:
refactor: Database password saved in secret

YkeLit committed:
refactor: note output is friendlier

YkeLit committed:
docs: Modify the field description in the README file

YkeLit committed:
test: change unittest

remove docker registry secret test
change test values

YkeLit committed:
chore: Update chart-testing-action version to v2.6.0 and add helm unittest plugin

fahchen commented on README.md:
```diff
@@ -0,0 +1,309 @@
+| secret.project_man_database_url | string | `"ecto://postgres:changeit@jet-postgresql/project_man_prod"` | Jet 数据库连接地址（格式：`ecto://USER:PASS@HOST/DATABASE`，[参考](https://hexdocs.pm/ecto/3.9.4/Ecto.Repo.html#module-urls)） |
+| secret.trace_aware_database_url | string | `"ecto://postgres:changeit@jet-postgresql/trace_aware_prod"` | Trace Aware 数据库连接地址（格式：`ecto://USER:PASS@HOST/DATABASE`，[参考](https://hexdocs.pm/ecto/3.9.4/Ecto.Repo.html#module-urls)） |
+| hosts | list | `[]` | jet 的访问地址 |
+| jetTLS | object | `{}` | tls 证书 |
```
jet 的前缀要么都保留要么都不要哈

fahchen commented on README.md:
````diff
@@ -0,0 +1,309 @@
+    tag: "latest"
+```
+
+### 2. 设置拉取镜像凭证
+
+```yaml
+imageCredentials:
+  # 设置仓库用户与密码
+  username: deploy-man@skylark
+  password: changeit
+  
+  # 使用已存在的仓库凭证
+  existImageSecrets: ""
+```
+
+> 镜像拉取凭证使用已有凭证与新建凭证设置一个即可
````
1. username, password 我们应该不用了吧
2. 这里做一个教程如何做一个 image credential 的 secret，或者贴一个教程的链接

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,229 @@
+hosts: []
+  # - localhost
+  # - jet.work
+jetTLS: {}
```
jet 的前缀是否需要，保持一致哈

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,229 @@
+    pullPolicy: IfNotPresent
+    tag: "latest"
+  env:
+    PROJECT_MAN_GRPC_SERVER_PORT: "{{ .Values.projectMan.service.port }}"
+    DYNAMIC_REPO_HOSTNAME: "{{ .Values.dynamicdb.fullnameOverride }}"
```
这几个用 template 取值的是不是直接写到 configmap 中，不允许通过 values 文件覆盖？

YkeLit replied:
这几个也是可以覆盖的

YkeLit replied:
这样写主要是可以将 service 中 port 的值与此处关联上

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,229 @@
+nameOverride: ""
+fullnameOverride: ""
+
+registry: registry.cn-hangzhou.aliyuncs.com
+existImageSecret: ""
```
```suggestion
image:
  registry: registry.cn-hangzhou.aliyuncs.com
  existingCredentialSecret: ""
```

fahchen commented on templates/secrets/environment.yaml:
```diff
@@ -0,0 +1,14 @@
+  name: {{ include "jet-helm-chart.fullname" . }}-env
+type: Opaque
+data:
+  credential_secret: {{ .Values.secret.credential_secret | b64enc }}
+  dynamic_repo_password: {{ .Values.secret.dynamic_repo_password | b64enc }}
+  jet_jwt_private_key: {{ .Values.secret.jet_jwt_private_key | b64enc }}
+  secret_key_base: {{ .Values.secret.secret_key_base | b64enc }}
+  trace_aware_database_url: {{ .Values.secret.trace_aware_database_url | b64enc }}
+  project_man_database_url: {{ .Values.secret.project_man_database_url | b64enc }}
```
这个地方从 values 中取出来的值，是不是还是在 values 中明文维护的呀？

YkeLit replied:
是的，这个地方手动去添加还没想好怎么弄

fahchen commented on templates/configmap/jet.yaml:
我觉得这个文件可以不存在，应该没有太多变量是共用的，如果共用可以在 env 那里设置

fahchen commented on templates/configmap/jet.yaml:
```diff
@@ -0,0 +1,13 @@
+    {{- include "jet-helm-chart.labels" . | nindent 4 }}
+  name: {{ include "jet-helm-chart.fullname" . }}-env
+data:
+  {{- tpl (.Values.env | toYaml | nindent 2) . }}
```
`.Values.env` 这个变量不存在哦

YkeLit replied:
存在的，在应用下面

YkeLit committed:
chore: fix unittest error

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,229 @@
+  # Custom tls certificate secret name
+  # tlsSecretName: ""
+
+# Jet subpath
+subpath: ""
+
+# Ingress middlewares
+middlewares:
+  corsSettings: {}
+    # accessControlAllowHeaders:
+    #   - '*'
+    # accessControlAllowMethods:
+    #   - GET
+    #   - OPTIONS
+    #   - PUT
+    # accessControlAllowOriginList:
+    #   - '*'
+    # accessControlMaxAge: "100"
```
给它们分个组呢

YkeLit replied:
之前说把 subpath 提到顶层的哇

fahchen commented on templates/breeze/deployment.yaml:
```diff
@@ -0,0 +1,85 @@
+            - '-g'
+            - '0.0.0.0:{{ .Values.breeze.service.port }}'
+          env:
+            - name: CREDENTIAL_SECRET
+              valueFrom:
+                secretKeyRef:
+                  name: {{ include "jet-helm-chart.fullname" . }}-env
+                  key: credential_secret
+            - name: DYNAMIC_REPO_PASSWORD
+              valueFrom:
+                secretKeyRef:
+                  name: {{ include "jet-helm-chart.fullname" . }}-env
+                  key: dynamic_repo_password
+            - name: JET_JWT_PRIVATE_KEY
+              valueFrom:
+                secretKeyRef:
+                  name: {{ include "jet-helm-chart.fullname" . }}-env
+                  key: jet_jwt_private_key
+            - name: SECRET_KEY_BASE
+              valueFrom:
+                secretKeyRef:
+                  name: {{ include "jet-helm-chart.fullname" . }}-env
+                  key: secret_key_base
+            - name: TRACE_AWARE_DATABASE_URL
+              valueFrom:
+                secretKeyRef:
+                  name: {{ include "jet-helm-chart.fullname" . }}-env
+                  key: trace_aware_database_url
+            - name: PROJECT_MAN_DATABASE_URL
+              valueFrom:
+                secretKeyRef:
+                  name: {{ include "jet-helm-chart.fullname" . }}-env
+                  key: project_man_database_url
```
这些的组装放对应的 env 那个 configmap 里面呢

fahchen commented on templates/project-man/ingressroute.yaml:
```diff
@@ -0,0 +1,69 @@
+{{- $webMiddlewares := list }}
+{{- if .Values.subpath }}
+{{- $webMiddlewares = append $webMiddlewares "strip-prefix" }}
+{{- end }}
+{{- if .Values.middlewares.corsSettings }}
+{{- $webMiddlewares = append $webMiddlewares "cors" }}
+{{- end }}
+{{- $webMiddlewares = append $webMiddlewares "compress" }}
+{{- if .Values.jetTLS }}
+{{- $webMiddlewares = append $webMiddlewares "redirect-to-https" }}
+{{- end }}
+
+{{/* Middlewares for websecure ingress-routes */}}
+{{- $websecureMiddlewares := list }}
+{{- if .Values.subpath }}
+{{- $websecureMiddlewares = append $websecureMiddlewares "strip-prefix" }}
+{{- end }}
+{{- if .Values.middlewares.corsSettings }}
+{{- $websecureMiddlewares = append $websecureMiddlewares "cors" }}
+{{- end }}
+{{- $websecureMiddlewares = append $websecureMiddlewares "compress" }}
```
这些是不是可以做一个 helper 来共享

fahchen commented on templates/trace-aware/deployment.yaml:
```diff
@@ -0,0 +1,93 @@
+      imagePullSecrets:
+        - name: {{ .Values.existImageSecrets }}
+      {{- end }}
+      initContainers:
+        {{- if .Values.projectmandb }}
+        - name: init-trace-aware-db
+          image: byzanteam/psql
+          env:
+            - name: PGPASSWORD
+              valueFrom:
+                secretKeyRef:
+                  name: jet-secret
+                  key: project-man-password
+          command:
+            - 'sh'
+            - '-c'
+            - 'until psql -c "SELECT 1" "user=postgres dbname=trace_aware_prod host={{ .Values.projectmandb.fullnameOverride }}"; do echo waiting for trace aware db; sleep 2; done'
+        {{- end }}
```
我看我们都写了 initContainer 来去检测对应的 service 有没有启动起来，如果不检测会出现什么问题呢？

fahchen commented on templates/_config-map.tpl:
这个文件名字改一下 config-map -> configmap

fahchen commented on templates/configmap/breeze.yaml:
要不你把 configmap 也维护到对应的文件夹下面吧

YkeLit committed:
refactor: Adjust the configmap directory and environment variable values

YkeLit committed:
refactor: remove secret values

use the existing secret as the environment variable

YkeLit committed:
refactor: add image field

YkeLit committed:
refactor: add ingressroute field, change  traefik.io/v1alpha1 api

YkeLit committed:
refactor: add ingressroute field, change  traefik.io/v1alpha1 api

YkeLit committed:
refactor: change jetTLS to certificate field, update README, templates, tests, and values

YkeLit committed:
docs: change image secrets info

YkeLit committed:
docs: add image pull certificate instructions

fahchen commented on templates/breeze/configmap.yaml:
```diff
@@ -0,0 +1,17 @@
+  {{- tpl (.Values.breeze.env | toYaml | nindent 2) . }}
+  BREEZE_GRPC_ADDRESS: {{ include "jet-helm-chart.breezeAddress" . }}
+  JET_API_ENDPOINT: {{ include "jet-helm-chart.jetApiEndpoint" . }}
+  DYNAMIC_REPO_USERNAME: postgres
+  DYNAMIC_REPO_EXPOSED_PORT: "5432"
```
这两个应该是从 dynamic db 那里来的引用值

YkeLit replied:
这两个需要做成动态的吗，username 没有引用值， port 可以引用

fahchen replied:
它们应该和数据库的设置保持一致

fahchen commented on templates/breeze/configmap.yaml:
```diff
@@ -0,0 +1,17 @@
+  PROJECT_MAN_GRPC_SERVER_PORT: "{{ .Values.projectMan.service.port }}"
+  TRACE_AWARE_GRPC_SERVER_PORT: "{{ .Values.traceAware.service.port }}"
+  DYNAMIC_REPO_HOSTNAME: "{{ .Values.dynamicdb.fullnameOverride }}"
+  DYNAMIC_REPO_DATABASE: dynamic_prod
```
这个也应该是引用值，你都检查一遍这个问题

YkeLit replied:
这个值没有应用，这是init 脚本创建的库名，所以在这个文件中写死了

fahchen replied:
如果 init 中改了，这里怎么同步呢？
或者如果用外部的数据库，这里又怎么同步呢？
思考把这些地方的变量做成一个来维护呢

YkeLit replied:
这个地方的话，我之前思考过，init 暂时没有办法和这里引用。目前我能想到的是把这个提到 values 文件这一层

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,218 @@
+  nodeSelector: {}
+  tolerations: []
+  affinity: {}
+  service:
+    name: project-man
+    type: ClusterIP
+    port: 50051
```
这个是不是没有设置的必要了？

YkeLit replied:
有啊，这个被 project man 的 configmap 引用

fahchen replied:
写到 values中就可以被覆盖哦，这个有没必要被覆盖呢？

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,218 @@
+      - credential_secret
+      - jet_jwt_private_key
+      - secret_key_base
+      - trace_aware_database_url
+      - project_man_database_url
```
这两个我们算不出来哇？

YkeLit replied:
目前算不出来

fahchen reviewed (commented):
大体上没有问题了

fahchen commented on README.md:
```diff
@@ -0,0 +1,312 @@
+| fullnameOverride | string | `""` |  |
+| nameOverride | string | `""` |  |
+| global.existingSecret.name | string | `jet-env-secret` |  |
+| global.existingSecret.keys[0] | string | `"credential_secret"` |  |
+| global.existingSecret.keys[1] | string | `"jet_jwt_private_key"` |  |
+| global.existingSecret.keys[2] | string | `"secret_key_base"` |  |
+| global.existingSecret.keys[3] | string | `"trace_aware_database_url"` |  |
+| global.existingSecret.keys[4] | string | `"project_man_database_url"` |  |
```
这几个 key 有什么用呀？

YkeLit replied:
这几个 key 是手动创建 secret 的 key

fahchen commented on README.md:
```diff
@@ -0,0 +1,312 @@
+| traceAware.service.name | string | `"trace-aware"` |  |
+| traceAware.service.port | int | `80` |  |
+| traceAware.service.type | string | `"ClusterIP"` |  |
+| secret.credential_secret | string | `""` | 用于加密应用数据库密码的密钥（`openssl rand -base64 48`） |
+| secret.dynamic_repo_password | string | `"changeit"` | 项目数据库密码 |
+| secret.jet_jwt_private_key | string | `""` | Jet 用于签名 JWT 的 RSA 私钥 |
+| secret.secret_key_base | string | `""` | Phoenix 用于生成会话密码的密钥 |
+| secret.project_man_database_url | string | `"ecto://postgres:changeit@jet-postgresql/project_man_prod"` | Jet 数据库连接地址（格式：`ecto://USER:PASS@HOST/DATABASE`，[参考](https://hexdocs.pm/ecto/3.9.4/Ecto.Repo.html#module-urls)） |
+| secret.trace_aware_database_url | string | `"ecto://postgres:changeit@jet-postgresql/trace_aware_prod"` | Trace Aware 数据库连接地址（格式：`ecto://USER:PASS@HOST/DATABASE`，[参考](https://hexdocs.pm/ecto/3.9.4/Ecto.Repo.html#module-urls)） |
```
这几个在 values 中没看到呢？

YkeLit committed:
feat: remove secret field in README file

YkeLit committed:
refactor: Add dynamicdb.database field and update configmap and values

YkeLit committed:
refactor: remove service field

YkeLit committed:
refactor: change database environment

YkeLit committed:
docs: Modify the field description in the README file

fahchen commented on templates/breeze/deployment.yaml:
```diff
@@ -29,18 +29,7 @@ spec:
-                secretKeyRef:
-                  name: {{ .Values.global.existingSecret.name }}
-                  key: dynamicdb-password
+            {{- include "jet-helm-chart.secretKeys" . | nindent 12 }}
```
这里不是所有的 secret 里面的 key 都被 breeze 需要的吧？

YkeLit replied:
那就需要给他拆开了

fahchen replied:
拆开是什么意思啊？

fahchen commented on templates/_deployment.tpl:
```diff
@@ -0,0 +1,20 @@
+- name: {{ $key | quote | replace "-" "_" | upper }}
+  valueFrom:
+    secretKeyRef:
+      name: jet-env-secret
```
这个 name 应该不是固定的吧

YkeLit replied:
name 直接给他固定下来比较好吧，从 README 到 资源文件都是这个名字就可以了

fahchen replied:
没必要固定吧，这个名字应该是安装这个 chart 的人提供的

YkeLit committed:
refactor: modify existingSecret.name usage and remove unused variables

fahchen commented on values.yaml:
```diff
@@ -0,0 +1,200 @@
+nameOverride: ""
+fullnameOverride: ""
+
+existingSecret:
+  name: jet-env-secret
```
这个为什么放一个 namespace 呢？现在这个 existingSecret 有点不清楚是什么的
可以就叫 existingJetSecret 哇

YkeLit replied:
放一个 name 的话，方便后面加其他参数，而不会影响到这个

fahchen replied:
其他的 secret 我看也没加这个东西，你举个可以例子呢

YkeLit replied:
比如我们现在只有 project man 用了 existing， 如果后面加应用，并且需要创建新的 secret 时，就可以在 exsitingJet Secret 下添加。或者我们后面还要添加一些 secret key 可以加 key 相关的变量

fahchen commented on templates/project-man/deployment.yaml:
```diff
@@ -0,0 +1,68 @@
+          image: {{ include "jet-helm-chart.projectManImage" . }}
+          imagePullPolicy: {{ .Values.projectMan.image.pullPolicy }}
+          env:
+            {{- include "jet-helm-chart.secretKeys" . | nindent 12 }}
```
我觉得不需要这个 helper 了，直接在这里写就可以了，因为只有这里用到了

YkeLit replied:
我给他移到 project man 目录下了

YkeLit committed:
refactor: Adjust the application of the env variable

YkeLit commented on templates/breeze/configmap.yaml:
@fahchen 这个要不要保留，因为现在没有使用

YkeLit commented on templates/breeze/deployment.yaml:
```diff
@@ -0,0 +1,57 @@
+            - '0.0.0.0:50052'
+            - '-r'
+            - '0.0.0.0:2137'
+          envFrom:
+            - configMapRef:
+                name: {{ include "jet-helm-chart.fullname" . }}-breeze-env
```
这里也是

YkeLit commented on values.yaml:
```diff
@@ -0,0 +1,196 @@
+
+projectMan:
+  image:
+    repository: jet/project_man
```
这里的镜像名要不要改成 jet，现在是 jet

YkeLit committed:
chore: change ct install action

YkeLit committed:
refactor: Remove unnecessary base64 decoding for secrets in lint-test.yaml

YkeLit committed:
refactor: Add `helm dependency update` step in lint-test.yaml to ensure accurate dependencies are installed

YkeLit committed:
refactor: Add charts directory to chart-dirs in ct.yaml for linting entire repository

YkeLit committed:
chore: Add bitnami Helm repository for dependency update

YkeLit committed:
chore: Reorder helm-extra-args in ct.yaml for consistency and clarity

YkeLit committed:
chore: Comment out setup-python action in lint-test.yaml

YkeLit committed:
chore: Update runs-on field in lint-test.yaml for consistency and stability

YkeLit committed:
docs: Update NOTES.txt to dynamically generate URLs with optional "https" based on presence of certificate field

YkeLit committed:
chore: Update helm-extra-args to helm-extra-set-args in ct.yaml for consistency and clarity

YkeLit committed:
chore: chart testing install test values file

YkeLit committed:
refactor: change secret values

YkeLit committed:
refactor: Update .helmignore to ignore .github/, .editorconfig, tests/ and ci/

fahchen commented on templates/project-man/deployment.yaml:
```diff
@@ -44,7 +44,7 @@ spec:
           imagePullPolicy: {{ .Values.projectMan.image.pullPolicy }}
           env:
-            {{- include "jet-helm-chart.secretKeys" . | nindent 12 }}
+            {{- include "jet-helm-chart.projectManSecretKeys" . | nindent 12 }}
```
project-man 这里应该也不是要所有的 secret keys 哦

YkeLit replied:
这里只有 project man 在用这个 helper，所以是所有的

fahchen replied:
projectMan 不会用全部的 key 吧

YkeLit replied:
![image](https://github.com/Byzanteam/jet-helm-chart/assets/48484963/8262fe48-e9a0-4eba-876d-9544136695e0)
这个 help 现在就只有这 5 个，全都是 project man 用的

fahchen replied:
trace-aware-database-url 呢？

YkeLit replied:
因为只有一个，我直接写到 deployment 的 env 里了

fahchen replied:
> project-man 这里应该也不是要所有的 secret keys 哦


> projectMan 不会用全部的 key 吧

所以 secret 中的东西不是全部给 projectMan 用的，所以你这里全部注入到 projectMan 中就不对了，projectMan 应该要选择自己要的

YkeLit replied:
![image](https://github.com/Byzanteam/jet-helm-chart/assets/48484963/ddc4a113-9e7e-4dff-b9de-78e485289364)
这里 project man 没有去加载 trace-aware-database-url  的 key 呀

YkeLit replied:
是要把 jet-env-secret 这个文件拆成两个吗，分别给 trace aware 和 project man

fahchen replied:
我没注意到这里😅

fahchen commented on .github/workflows/lint-test.yaml:
```diff
@@ -0,0 +1,49 @@
+      - name: Create kind cluster
+        uses: helm/kind-action@v1.4.0
+
+      - name: Install dependent environment
+        run: |
+          helm repo add traefik https://traefik.github.io/charts
+          helm install traefik traefik/traefik
+          helm repo add bitnami https://charts.bitnami.com/bitnami
+          helm dependency update
+          echo -n "${{ secrets.DOCKER_REGISTRY_SECRET }}" | kubectl apply -n default -f -
+          echo -n "${{ secrets.JET_ENV_SECRET }}" | kubectl apply -n default -f -
+          echo -n "${{ secrets.DEPENDENCY_DB_SECRET }}" | kubectl -n default apply -f -
+
+      - name: Run chart-testing (install)
+        run: ct install --config .github/linters/ct.yaml
```
用 ct install 测试它会真实地拉取镜像和连接数据库吗？它是怎么判断测试通过的呢？

YkeLit replied:
是会去拉取镜像和链接数据库的，https://github.com/helm/chart-testing/blob/1f5cd2003aed452dd324bd14b4a790b38b07b346/pkg/chart/chart.go#L517-L541
代码注释给了说明

fahchen commented:
不用一直拉我的 review
1. 你正常回复我可以收到提醒
2. 如果本次只有代码提交，改动到位了，就拉我的 review

YkeLit commented:
> 不用一直拉我的 review
> 
> 1. 你正常回复我可以收到提醒
> 2. 如果本次只有代码提交，改动到位了，就拉我的 review

好

fahchen commented on templates/trace-aware/deployment.yaml:
```diff
@@ -0,0 +1,66 @@
+            - name: PGPASSWORD
+              valueFrom:
+                secretKeyRef:
+                  name: dependency-db-secret
+                  key: project-man-password
```
如果使用外部的数据库，这里的东西还有问题

YkeLit replied:
如果使用外部数据库，这里不会创建 initcontainer

fahchen reviewed (approved):
🧡💙💚💛💜❤️
